sslr-core/src/main/java/com/sonar/sslr/impl/typed/SyntaxTreeCreator.java

```diff
@@ -0,0 +1,191 @@
+      } else {
+        Object child = visit(node.getChildren().get(0));
+        if (child instanceof AstNode) {
+          ((AstNode) child).hasToBeSkippedFromAst();
+        }
+        return Optional.of(child);
+      }
+    }
+
+    List<ParseNode> children = node.getChildren();
+    List<Object> convertedChildren = Lists.newArrayList();
+    for (ParseNode child : children) {
+      Object result = visit(child);
+
+      if (result != null) {
+        // FIXME to remove aafter full migration: Allow to skip optional nodes that are supposed to bw skipped from the AST
+        if ((result instanceof Optional && ((Optional) result).isPresent() && ((Optional) result).get() instanceof AstNode) && ((AstNode) ((Optional) result).get()).hasToBeSkippedFromAst()){
+          for (AstNode resultChild : ((AstNode) ((Optional) result).get()).getChildren()) {
+            convertedChildren.add(resultChild);
+          }
+
+        } else if (result instanceof AstNode && ((AstNode) result).hasToBeSkippedFromAst()) {
+          for (AstNode resultChild : ((AstNode) result).getChildren()) {
+            convertedChildren.add(resultChild);
+          }
+        } else {
+          convertedChildren.add(result);
+        }
+      }
+    }
+
+    if (mapping.isOneOrMoreRule(ruleKey)) {
+      return Lists.newArrayList(convertedChildren);
+    }
+
+    if (mapping.isZeroOrMoreRule(ruleKey)) {
+      return convertedChildren.isEmpty() ? Optional.absent() : Optional.of(Lists.newArrayList(convertedChildren));
+    }
+
+    Method method = mapping.actionForRuleKey(ruleKey);
+    if (method == null) {
+      int startIndex = node.getStartIndex();
+      int endIndex = node.getEndIndex();
+
+      return nodeBuilder.createNonTerminal(rule.getRuleKey(), rule, convertedChildren, startIndex, endIndex);
+    }
+
+    try {
+      return method.invoke(treeFactory, convertedChildren.toArray(new Object[convertedChildren.size()]));
+    } catch (IllegalAccessException e) {
+      throw Throwables.propagate(e);
+    } catch (IllegalArgumentException e) {
+      throw Throwables.propagate(e);
+    } catch (InvocationTargetException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private Object visitTerminal(ParseNode node) {
+    TokenType type = null;
+    if (node.getMatcher() instanceof TriviaExpression) {
+      TriviaExpression ruleMatcher = (TriviaExpression) node.getMatcher();
+      if (ruleMatcher.getTriviaKind() == TriviaKind.SKIPPED_TEXT) {
+        return null;
+      } else if (ruleMatcher.getTriviaKind() == TriviaKind.COMMENT) {
+        addComment(node);
+        return null;
+      } else {
+        throw new IllegalStateException("Unexpected trivia kind: " + ruleMatcher.getTriviaKind());
+      }
+    } else if (node.getMatcher() instanceof TokenExpression) {
+      TokenExpression ruleMatcher = (TokenExpression) node.getMatcher();
+      type = ruleMatcher.getTokenType();
+      if (GenericTokenType.COMMENT == ruleMatcher.getTokenType()) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)